### PR TITLE
feat: 'kpm add' supports add oci registry packages.

### LINKED ISSUE #74: feature: simplify command `kpm add`, the command is too long.
## Feature Request

**Is your feature request related to a problem? Please describe:**

`kpm add` currently only supports adding kcl packages dependency in the git repository.

And the command of `kpm add` is too long for user to type it in shell, which creates a bad user experience.

```
kpm add -git https://github.com/awesome-kusion/konfig.git -tag v0.0.1
```

**Describe the feature you'd like:**

If I want to import a k8s kcl package to develop a kcl program.

1. Add the kcl package named `k8s`.
```
# latest version will be used
kpm add k8s
```

or add the kcl package named `k8s` with version `v0.1.0`.
```
# version 0.1.0 will be used
kpm add k8s:0.1.0
```

2. Import the k8s package in `main.k`.

```kcl
import k8s.api.core.v1 as k8core

nginx = k8core.Pod {
    metadata.name = "web-app"
    spec.containers = [{
        name = "main-container"
        image = "nginx"
        ports = [{containerPort = 80}]
    }]
}
```

3. Compile it.
```
kpm run
```

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by zong-zhe**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

fix #74 

#### 2. What is the scope of this PR (e.g. component or file name):

kpm add

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [x] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/settings/settings.go [hunk 1]
```diff
@@ -1,36 +1,135 @@
...
-	return filepath.Join(home, CONFIG_JSON_PATH), nil
+	return filepath.Join(home, jsonFileName), nil
 }
 
-// Init returns default kpm settings.
+// Init returns default kpm settings load from '$KCL_PKG_PATH/.kpm/config/kpm.json'
+// and '$KCL_PKG_PATH/.kpm/config/config.json'.
 func Init() (*Settings, error) {
-	credentialsFile, err := GetConfigJsonPath()
+	if kpm_settings == nil {
```
[the remark this answers is not in the document]
Comment (Peefy) [reply pkg/settings/settings.go:69]: > Note that there may be global concurrency issues here, which require locking using methods such as' sync 'and' init 'functions.

Besides, all places where files are written will have similar issues. For example, kcl. mod. lock, etc.

File: pkg/cmd/cmd_add.go [hunk 2]
```diff
@@ -99,17 +85,93 @@ func onlyOnceOption(c *cli.Context, name string) (*string, error) {
...
+			LocalPath:    localPath,
+			RegistryOpts: *ociReg,
+		}, nil
+	}
+
+}
+
+// parseGitRegistryOptions will parse the git registry information from user cli inputs.
+func parseGitRegistryOptions(c *cli.Context) (*opt.RegistryOptions, error) {
+	gitUrl, err := onlyOnceOption(c, "git")
```
Comment (Peefy) [inline pkg/cmd/cmd_add.go:120]: The feeling here is inconsistent with the OCI's options parse process. Ref:  https://buf.build/docs/tutorials/getting-started-with-bsr/

Comment (Peefy) [reply pkg/cmd/cmd_add.go:120]: > The feeling here is inconsistent with the OCI's options parse process. Ref: https://buf.build/docs/tutorials/getting-started-with-bsr/

Do we necessarily need git and tag options?

File: pkg/mod/modfile.go [hunk 3]
```diff
@@ -54,27 +69,50 @@ type Dependency struct {
...
+			return err
+		}
+		dep.Source.Oci.Reg = settings.DefaultOciRegistry()
+		dep.Source.Oci.Repo = filepath.Join(settings.DefaultOciRepo(), dep.Name)
+	}
+	return nil
+}
+
 // Download will download the kcl package to localPath from registory.
 func (dep *Dependency) Download(localPath string) (*Dependency, error) {
```
Comment (Peefy) [inline pkg/mod/modfile.go:86]: Does the Download function support concurrency?

File: pkg/opt/opt.go [hunk 4]
```diff
@@ -63,25 +65,20 @@ func (opts *GitOptions) Validate() error {
...
 
-func GetOCIReg() string {
-	return DEFAULT_REGISTRY
-}
-
 func GetDefaultOCITag() string {
 	return DEFAULT_OCI_TAG
 }
 
 type OciOptions struct {
```
Comment (Peefy) [inline pkg/opt/opt.go:74]: Add some comments for `OciOptions`: such as {reg}/{repo}/{pkg}/{tag}

File: pkg/settings/settings.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zong-zhe) [reply pkg/settings/settings.go]: fixed. I used `sync.Once` to enhancement the singleton settings. And `flock` to lock the `kcl.mod.lock` before `kpm add`.

File: pkg/opt/opt.go (hunk 4 above)
Comment (zong-zhe) [reply pkg/opt/opt.go:76]: fixed.

File: pkg/cmd/cmd_add.go (hunk 2 above)
Comment (zong-zhe) [reply pkg/cmd/cmd_add.go:141]: This implementation references the cargo command, which can be adjusted later to separate the git related content from the OCI registry into two different commands. Or if adding a git repository as a dependency is not needed, git related content can be removed later.

File: pkg/cmd/cmd_add.go [hunk 5]
```diff
@@ -99,17 +85,93 @@ func onlyOnceOption(c *cli.Context, name string) (*string, error) {
...
 }
 
-func addGitDep(opt *opt.AddOptions, kclPkg *pkg.KclPkg) error {
-	if opt.RegistryOpts.Git == nil {
-		reporter.Report("kpm: a value is required for '-git <URI>' but none was supplied")
-		reporter.ExitWithReport("kpm: run 'kpm add help' for more information.")
+// parseAddOptions will parse the user cli inputs.
+func parseAddOptions(c *cli.Context, localPath string) (*opt.AddOptions, error) {
+	// parse from 'kpm add -git https://xxx/xxx.git -tag v0.0.1'.
+	if c.NArg() == 0 {
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply pkg/cmd/cmd_add.go:117]: For extension, If the expansion scenario is different from OCI, maybe kpm can add new subcommands to achieve it.

File: pkg/cmd/cmd_add.go [hunk 6]
```diff
@@ -99,17 +107,92 @@ func onlyOnceOption(c *cli.Context, name string) (*string, error) {
...
+// parseOciPkgNameAndVersion will parse package name and version
+// from string "<pkg_name>:<pkg_version>".
+func parseOciPkgNameAndVersion(s string) (string, string) {
+	parts := strings.Split(s, ":")
+	if len(parts) == 1 {
+		return parts[0], ""
+	}
+
+	if len(parts) > 2 {
+		return "", ""
```
Comment (Peefy) [inline pkg/cmd/cmd_add.go:194]: Do we need to return an err here?

File: pkg/settings/settings.go [hunk 7]
```diff
@@ -1,36 +1,126 @@
...
 	if err != nil {
 		return nil, err
 	}
-	return &Settings{
-		CredentialsFile: credentialsFile,
-	}, nil
+
+	defaultKpmConf := DefaultKpmConf()
+
+	b, err := ioutil.ReadFile(kpmConfpath)
```
Comment (Peefy) [inline pkg/settings/settings.go:101]: os.ReadFile

File: pkg/settings/settings.go [hunk 8]
```diff
@@ -1,36 +1,126 @@
...
+		err = os.MkdirAll(filepath.Dir(kpmConfpath), 0755)
+		if err != nil {
+			return nil, err
+		}
+
+		b, err := json.Marshal(defaultKpmConf)
+		if err != nil {
+			return nil, err
+		}
+		err = ioutil.WriteFile(kpmConfpath, b, 0644)
```
Comment (Peefy) [inline pkg/settings/settings.go:114]: os.WriteFile

Comment (Peefy) [reply pkg/settings/settings.go:114]: > os.WriteFile

https://github.com/helm/helm/blob/main/internal/fileutil/fileutil.go AtomicWriteFile

File: pkg/settings/settings.go [hunk 9]
```diff
@@ -1,36 +1,126 @@
...
 
-	return filepath.Join(home, CONFIG_JSON_PATH), nil
+	return filepath.Join(home, jsonFileName), nil
+}
+
+// GetSettings will return the kpm setting singleton.
+func GetSettings() (*Settings, error) {
+	var err error
+	once.Do(func() {
+		credentialsFile, err := GetFullJsonPath(CONFIG_JSON_PATH)
```
Comment (Peefy) [inline pkg/settings/settings.go:72]: Ref: https://github.com/helm/helm/blob/main/pkg/helmpath/lazypath.go

Comment (zong-zhe) [reply pkg/settings/settings.go:71]: fixed.

File: pkg/settings/settings.go (hunk 8 above)
Comment (zong-zhe) [reply pkg/settings/settings.go]: Currently, kpm does not support concurrent environments, and kpm will support concurrent environments in future. This section can be adjusted in the process of enhancing concurrency in kpm.

File: pkg/settings/settings.go (hunk 7 above)
Comment (zong-zhe) [reply pkg/settings/settings.go]: fixed.

File: pkg/cmd/cmd_add.go (hunk 6 above)
Comment (zong-zhe) [reply pkg/cmd/cmd_add.go]: fixed.

File: pkg/cmd/cmd_add.go [hunk 10]
```diff
@@ -99,17 +107,96 @@ func onlyOnceOption(c *cli.Context, name string) (*string, error) {
...
+	if len(parts) == 1 {
+		return parts[0], "", nil
+	}
+
+	if len(parts) > 2 {
+		return "", "", errors.InvalidAddOptionsInvalidOciRef
 	}
 
-	return kclPkg.AddDeps(opt)
+	return parts[0], parts[1], nil
```
Comment (Peefy) [inline pkg/cmd/cmd_add.go:201]: Considering the invalid format: `kusionstack.io/kclvm:`

Comment (Peefy) [reply pkg/cmd/cmd_add.go:201]: > Considering the invalid format: `kusionstack.io/kclvm:`

You can refer to some standard implementations.

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**